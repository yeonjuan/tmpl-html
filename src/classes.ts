import {isStringLike, } from "./utils";

type Value = string | number | boolean | undefined | null;
type RecordValue = Record<string, unknown>;
type ArrayValue = Arg[];
type Arg = Value | RecordValue | ArrayValue;

const hawOwn = {}.hasOwnProperty;

export default function classes(...args: Arg[]): string {
  const classnames: string[] = [];

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (isStringLike(arg)) {
      classnames.push(String(arg));
    } else if (Array.isArray(arg)) {
      arg && classes(arg);
    } else if (typeof arg === 'object' && !!arg) {
      for (const key in arg) {
        (arg[key] && hawOwn.call(arg, key) && classnames.push(key))
      }
    }
  }
  return `"${classnames.join(" ")}"`;
}
