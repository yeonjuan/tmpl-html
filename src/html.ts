import {isNumeric} from './utils';
import convert from './convert';
import type {StringLikeType} from './types';

export default function html(
  strings: TemplateStringsArray,
  ...args: unknown[]
) {
  if (!args || args.length <= 0) {
    return strings.join("");
  }
  const result: StringLikeType[] = [];
  for (let i = 0; i < strings.length; i++) {
    result.push(strings[i]);
    const arg = args[i] || isNumeric(args[i]) ? (args[i] as string) : "";
    arg && result.push(convert(arg));
  }
  return result.join("");
}
