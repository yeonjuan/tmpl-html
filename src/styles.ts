import {hasOwn} from './utils';

type Value = string | number | boolean | undefined | null;
type RecordValue = Record<string, Value>;

export default function styles(...args: RecordValue[]): string {
  const styleMap: any = {};
  const style: string[] = [];

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (typeof arg === 'object' && !!arg) {
      for (const key in arg) {
        if (arg[key] && hasOwn.call(arg, key)) {
          styleMap[key] = arg[key];
        }
      }
    }
  }

  for (const key in styleMap) {
    if (key && styleMap[key] && hasOwn.call(styleMap, key)) {
      style.push(`${key}:${styleMap[key]}`);
    }
  }

  return `"${style.join(';')}"`;
}
