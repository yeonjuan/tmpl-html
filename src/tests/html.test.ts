import classes from "../classes";
import html from "../html";
import styles from "../styles";

test.each([
  () => html`<div></div>`,
  () => {
    const child = html`test`;
    return html`<div>${child}</div>`;
  },
  () => {
    const children = Array.from({length: 5}, (_, index) => html`<li>${index}</li>`);
    return html`
      <ul>${children}</ul>
    `;
  },
  () => {
    const text = 'text';
    return html`<div>${text}</div>`;
  },
  () => {
    const style = {
      width: '100px',
      height: '50px'
    }
    return html`
      <div style=${styles(style)} class=${classes('a', 'b')}></div>
    `
  }
])('html', (func) => {
  expect(func()).toMatchSnapshot()
});
