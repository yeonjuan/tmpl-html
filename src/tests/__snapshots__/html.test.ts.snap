// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`html 1`] = `"<div></div>"`;

exports[`html 2`] = `"<div>test</div>"`;

exports[`html 3`] = `
"
      <ul><li></li><li></li><li></li><li></li><li></li></ul>
    "
`;

exports[`html 4`] = `"<div>text</div>"`;

exports[`html 5`] = `
"
      <div style=\\"width:100px;height:50px\\" class=\\"a b\\"></div>
    "
`;
